src/main/java/org/jenkinsci/remoting/engine/JnlpConnectionStateListener.java

```diff
@@ -0,0 +1,111 @@
+package org.jenkinsci.remoting.engine;
+
+import hudson.remoting.Channel;
+import hudson.remoting.ChannelBuilder;
+import java.net.Socket;
+import javax.annotation.Nonnull;
+import org.jenkinsci.remoting.protocol.impl.ConnectionRefusalException;
+
+/**
+ * A listener for connection state changes.
+ *
+ * The listeners will be called in order for any unique event. Use
+ * {@link JnlpConnectionState#setStash(JnlpConnectionState.ListenerState)} and
+ * {@link JnlpConnectionState#getStash(Class)} if you need to maintain state between callbacks.
+ *
+ * <ol>
+ * <li>{@link #beforeProperties(JnlpConnectionState)} (if closed then {@link #afterDisconnect(JnlpConnectionState)})
+ * </li>
+ * <li>{@link #afterProperties(JnlpConnectionState)} (if closed then {@link #afterDisconnect(JnlpConnectionState)})</li>
+ * <li>{@link #beforeChannel(JnlpConnectionState)} (if closed then {@link #afterDisconnect(JnlpConnectionState)})</li>
+ * <li>{@link #afterChannel(JnlpConnectionState)} (if closed then {@link #channelClosed(JnlpConnectionState)})</li>
+ * <li>{@link #channelClosed(JnlpConnectionState)}</li>
+ * <li>{@link #afterDisconnect(JnlpConnectionState)}</li>
+ * </ol>
+ *
+ * @since FIXME
+ */
+public abstract class JnlpConnectionStateListener {
+    /**
+     * Notification that the connection has been established and properties will be exchanged.
+     * Call {@link JnlpConnectionState#ignore()} to suppress any further notifications of this event.
+     * Call {@link JnlpConnectionState#reject(ConnectionRefusalException)} to reject the connection.
+     * Call {@link JnlpConnectionState#approve()} to declare ownership of this event (normally better to defer this to
+     * {@link #afterProperties(JnlpConnectionState)})
+     *
+     * @param event the event.
+     */
+    public void beforeProperties(@Nonnull JnlpConnectionState event) {
+
+    }
+
+    /**
+     * Notification that properties have been exchanged.
+     * Call {@link JnlpConnectionState#ignore()} to suppress any further notifications of this event.
+     * Call {@link JnlpConnectionState#reject(ConnectionRefusalException)} to reject the connection.
+     * Call {@link JnlpConnectionState#approve()} to declare ownership of this event.
+     *
+     * @param event the event.
+     */
+    public abstract void afterProperties(@Nonnull JnlpConnectionState event);
+
+    /**
+     * Callback to allow the {@link JnlpConnectionState#approve()} listener to decorate the {@link ChannelBuilder} via
+     * {@link JnlpConnectionState#getChannelBuilder()}.
+     *
+     * @param event the event.
+     */
+    public void beforeChannel(@Nonnull JnlpConnectionState event) {
+    }
+
+    /**
+     * Callback to notify the {@link JnlpConnectionState#approve()} listener that the {@link Channel} has been
+     * created and is available via {@link JnlpConnectionState#getChannel()}.
+     *
+     * @param event the event.
+     */
+    public abstract void afterChannel(@Nonnull JnlpConnectionState event);
```

Comment:
That is how you pass the channel into the Node or whatever else (*cough* Shared Slave *cough* Shared JNLP Cloud) that needs to receive it